app/models/user.rb

```diff
@@ -3,11 +3,32 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
-
+  devise :omniauthable, omniauth_providers: [:facebook]
   belongs_to :company, optional: true
   has_many :messages
   # has_many :profiles // REPLACED BY BELOW METHOD 'PROFILES'
 
+  def self.find_for_facebook_oauth(auth)
+    user_params = auth.slice("provider", "uid")
+    user_params.merge! auth.info.slice("email", "first_name", "last_name")
+    user_params[:facebook_picture_url] = auth.info.image
+    user_params[:token] = auth.credentials.token
+    user_params[:token_expiry] = Time.at(auth.credentials.expires_at)
+    user_params = user_params.to_h
+
+    user = User.find_by(provider: auth.provider, uid: auth.uid)
+    user ||= User.find_by(email: auth.info.email) # User did a regular sign up in the past.
+    if user
+      user.update(user_params)
+    else
+      user = User.new(user_params)
+      user.password = Devise.friendly_token[0, 20] # Fake password for validation
+      user.save
+    end
+
+    return user
+  end
+
```

Comment:
Is it advisable to leave this in the model, or should this be put in a helper? Don't think it will make any difference, but for organisation Marin already advised us to put the euro_thousand_separate method in a helper called "format_money.rb".

Reply:
@Danrod16 Actually I take the above back... it was only because we didn't have a money model